applications/dashboard/models/class.importmodel.php

```diff
@@ -2228,6 +2228,7 @@ public function loadTableInsert($tablename, $path, $skipHeader = true, $chunk =
                 $inserts .= ',';
             }
             $inserts .= '('.implode(',', $row).')';
+            $inserts = str_replace("'NULL',", "null,", $inserts);  // Need to replace any 'NULL',  values with actual null value for database
```

Comment:
Was it ever tested? 
IMHO this line will do nothing. because `$inserts` string should contain string like `(var1,var2,var3,NULL,var4)`
no quotes until quotes (single or double) is a part of those variable values.

Reply:
You have to look at the fGetCSV2 function when it reads in the txt files with the data in them they could have \N to represent NULL value. when it reads in the file it properly creates an array of values for each filed form the txt file and the \N get converted to null properly in the array. Problem is when it is returned to the loadTableInsert an trim is done on the Array and it changes all the NULL values to.. so to get aroudn this i suggest we change \N to the string value 'NULL' then after the trim and implode we just replace the 'NULL', with   null, this was the only way I could get the process to function and to get my data imported into VB

Next to that just replace the subroutine in fGetCSV2 that replaces the \N with null and jsut leave the \N in there till just before the inset and replace \N with null

Reply:
Here is an example of an export that would just error out on import

```
RoleID:int,Garden.SignIn.Allow:tinyint,ForumPermissions:varchar(255),GenericPermissions:varchar(255),Garden.Profiles.View:tinyint,Garden.Activity.View:tinyint,Garden.Profiles.Edit:tinyint,Plugins.Signatures.Edit:tinyint(1),Vanilla.Discussions.View:tinyint,Vanilla.Discussions.Add:tinyint,Vanilla.Comments.Add:tinyint,Plugins.Attachments.Download:tinyint(1),Plugins.Attachments.Upload:tinyint(1),Garden.Moderation.Manage:tinyint,Vanilla.Discussions.Announce:tinyint,Vanilla.Discussions.Close:tinyint,Vanilla.Discussions.Delete:tinyint,Vanilla.Comments.Delete:tinyint,Vanilla.Comments.Edit:tinyint,Vanilla.Discussions.Edit:tinyint,Vanilla.Discussions.Sink:tinyint,Garden.Activity.Delete:tinyint,Garden.Users.Add:tinyint,Garden.Users.Edit:tinyint,Garden.Users.Approve:tinyint,Garden.Users.Delete:tinyint,Garden.Applicants.Manage:tinyint(1),Garden.Settings.View:tinyint,Garden.Settings.Manage:tinyint,Garden.Messages.Manage:tinyint(1),Vanilla.Spam.Manage:tinyint(1)
"1",0,"9048079","64",\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
```

Reply:
@MrCaspan It is clear what you are trying to achieve. My comment only relates to one particular line of code.
From what I can understand before your changes we have some array like:
```
["1",0,"9048079","64",\N,\N, ...]
```
you want to transform it to
```
["1",0,"9048079","64","NULLL","NULLL", ...]
```
and then  we implode it to:
```
(1,0,9048079,64,NULLL,NULLL, ...)
```
please notice that imploded string has no quotes!
https://ideone.com/i8HuS7

That means that you can't replace with `str_replace("'NULL',"`
because there is no such substring there

Reply:
that's almost correct except \N get read in as a string value so your first array shoudl be  
`["1",0,"9048079","64","\N","\N", ...]`
Then I replace the "\N" with "NULL"
Then just before the insert replace "NULL", with NULL,   i check for the comma in the text string just to make sure i have captured a NULL value and not someone typing NULL in a comment or something

Next to that you could not replace the \N in the fGetCSV2 function but leave it there till just after the implode and replace "\N", will NULL, at that point.. one less step.

Reply:
Good point on your implode.. let me double check that!!

Reply:
okay now i know why I had to do this... there is a subroutine in fGetCSV2 that takes all the Array elements and puts them in '' so that when its imploded there is still single quotes around text So in the array you will see NULL will be [8] => "'NULL'" that is why i have to search for 'NULL', and replace it with just NULL,

Reply:
I believe one of the problems is the line trimming the row values:
`$row = array_map('trim', $row);`.
This will transform `NULL` values to `''` strings.
Casting these _NULL_ values to a _'null'_ string before trimming will circumvent this issue. Like so:

```suggestion
           foreach ($row as &$col_value) {
                if (is_null($col_value)) {
                    $col_value = 'null';
                }
            }
            $row = array_map('trim', $row);
```
and
```suggestion
            $inserts = str_replace("'null'", 'NULL', $inserts);
```